_posts/2017-10-27-conventional-interfaces.md

```diff
@@ -0,0 +1,225 @@
+
+
+Most of the ideas and some of the code samples contained in this post originally come from the SICP book. Although it is written using Lisp the ideas and concepts are applicable to most languages. We will be following them in **Haskell** as it is the language I am currently learning and enjoying, its terse syntax also makes it a great fit for a blog post. 
+
+## What are Conventional Interfaces
+
+SICP describes conventional interfaces as a design principle for working with data structures. It is composed of a set of standard operators or combinators that connect the different steps required to implement computations in computer programs. 
+
+The combinators in question may be familiar to you as they will probably be provided in some form in your preferred language, e.g. `map`, ``filter``, ``flatMap`` (also now as ``bind``), `reduce` (also known as `fold`). They let us capture common patterns in the implementation of programs that a priori are structurally different, enabling us to think and reason about different programs in the same way. This is of great importance as it enables us to intuitively express completely different and unrelated programs applying function composition to this set of combinators.
+
+SCIP also introduces the concept of signal-processing as a metaphor to reason about this programming style.
+
+> A signal-processing engineer would find it natural to conceptualize processes in terms of signals flowing through a cascade of stages, each of which implements part of the program plan. 
+
+We will expand more on this metaphor later on, but for now, it is important to emphasize that making the signal-flow structure evident in the design of our programs increases the modularity and readability of the resulting code. It is also important to emphasize that the use of these combinators increase the level of abstraction at which we can write code by taking away low-level operations such as iteration, recursion or conditional statements.
+
+To demonstrate the importance of the principle, let's consider the two following functions taken from SICP, which I have translated from Lisp to Haskell.
+
+*A simple type representing a binary tree*
+
+```
+data BinaryTree a =
+    Leaf a
+  | Node (BinaryTree a) (BinaryTree a) deriving (Foldable)
+```
+
+*Sum squares of the leaves of a tree that are odd*
+
+
+```
+sumOddSquares :: (Integral a) => BinaryTree a -> a
+sumOddSquares (Node l r)           = sumOddSquares l + sumOddSquares r
+sumOddSquares (Leaf x) | odd x     = x ^ 2
+                       | otherwise = 0
+```
+
+*All the even Fibonacci numbers up to a given number*
```

Comment:
"All the even Fibonacci numbers up to a given number"
Should it be "Sum of all the ..." ?

Reply:
This is not summing them is returning the list with all the even fib numbers

Reply:
Ooopsies